From bd8c325b9cd2d46a5a0f81e1a537ba95be52e442 Mon Sep 17 00:00:00 2001
From: Patrick Rudolph <patrick.rudolph@9elements.com>
Date: Thu, 12 Mar 2026 16:04:25 +0100
Subject: [PATCH] overlay: Fix "Support MuxChannel object for dynamic buses"

entity-manger crashes in various places and shows strange behaviours.

DeviceConfig::buildDevice used a pointer instead of a reference to
self, causing random crashed when the object was destroyed before
the timer expired. Use a weak_ptr to prevent the crash and random
behaviour.

Upstream-Status: Pending
Signed-off-by: Patrick Rudolph <patrick.rudolph@9elements.com>
---
 src/entity_manager/overlay.cpp | 11 +++++++++--
 src/entity_manager/overlay.hpp |  3 ++-
 2 files changed, 11 insertions(+), 3 deletions(-)

diff --git a/src/entity_manager/overlay.cpp b/src/entity_manager/overlay.cpp
index 519a84b..3b69e65 100644
--- a/src/entity_manager/overlay.cpp
+++ b/src/entity_manager/overlay.cpp
@@ -173,19 +173,26 @@ bool DeviceConfig::buildDevice(boost::asio::io_context& io,
         if (!createDevice())
         {
             deleteDevice();
+            std::weak_ptr<DeviceConfig> weakRef = weak_from_this();
 
             std::shared_ptr<boost::asio::steady_timer> createTimer =
                 std::make_shared<boost::asio::steady_timer>(io);
             createTimer->expires_after(std::chrono::milliseconds(500));
             createTimer->async_wait(
-                [createTimer, this, retries,
+                [createTimer, weakRef, retries,
                  &io](const boost::system::error_code& ec) mutable {
                     if (ec)
                     {
                         lg2::error("Timer error {ERR}", "ERR", ec.message());
                         return false;
                     }
-                    return buildDevice(io, retries - 1);
+                    std::shared_ptr<DeviceConfig> self = weakRef.lock();
+
+                    if (self) {
+                        return self->buildDevice(io, retries - 1);
+                    } else {
+                        return false;
+                    }
                 });
             return false;
         }
diff --git a/src/entity_manager/overlay.hpp b/src/entity_manager/overlay.hpp
index 27e5341..5d79d0f 100644
--- a/src/entity_manager/overlay.hpp
+++ b/src/entity_manager/overlay.hpp
@@ -19,7 +19,8 @@ constexpr const char* i2CDevsDir = "/sys/bus/i2c/devices";
 constexpr const char* muxSymlinkDir = "/dev/i2c-mux";
 const std::regex illegalNameRegex("[^A-Za-z0-9_]");
 
-class DeviceConfig
+class DeviceConfig :
+    public std::enable_shared_from_this<DeviceConfig>
 {
   public:
     DeviceConfig() = delete;
-- 
2.47.1

